### PR TITLE
convert `cpapi_artifact_download` to use dropshot / progenitor streaming interfaces

### PR DESCRIPTION
**Pull request by ahl**
This lets us turn on `test_write_artifact_to_filesystem` which was previously ignored. I hope this is a good thing.

### CODE REVIEW

File: sled-agent/src/updates.rs [hunk 1]
```diff
@@ -48,24 +46,27 @@ pub async fn download_artifact(
...
+            // TODO it would be better to stream this into the file rather than
+            // accumulating it in memory.
+            let contents = response
+                .into_inner_stream()
+                .try_fold(Vec::new(), |mut acc, x| async move {
+                    acc.extend(x);
+                    Ok(acc)
+                })
+                .await
+                .map_err(|e| Error::Response(e.into()))?;
```
Comment (iliana) [inline sled-agent/src/updates.rs:68]: This should be easier to stream now if we want, e.g.

```rust
let mut file = tokio::fs::File::open(&tmp_path).await?;
let mut stream = response.into_inner_stream();
while let Some(chunk) = stream.next().await {
    file.write_all(&chunk?);
}
```

(adding `.map_err` as appropriate). Alternately you could do something similar with `.try_for_each` on the stream.

**Review by iliana — APPROVED** [review]
Yes!!

File: sled-agent/src/updates.rs (hunk 1 above)
Comment (ahl) [reply sled-agent/src/updates.rs:68]: so I played around with that a bit, but didn't land on something I loved. I'll try again as you suggest.

What should we be doing if we encounter an error part way through writing the file? I don't see us removing that temp file.

Comment (ahl) [reply sled-agent/src/updates.rs]: Ok, I couldn't make `try_for_each` work... I don't think I understand how to use `file` from within the closure properly. But I used a `while` loop as you suggested.

Comment (iliana) [reply sled-agent/src/updates.rs]: > What should we be doing if we encounter an error part way through writing the file? I don't see us removing that temp file.

Hm. I suppose we could use `tempfile` here... I have some other code for a side project that is using that combined with tokio::fs::File. I'll push something on top of this branch unless you want to give it a shot?

I used [`NamedTempFile::into_parts`](https://docs.rs/tempfile/latest/tempfile/struct.NamedTempFile.html#method.into_parts) and then [`tokio::fs::File::from_std`](https://docs.rs/tokio/latest/tokio/fs/struct.File.html#method.from_std). Once the `TempPath` from `into_parts` gets dropped the file is cleaned up.

Comment (ahl) [reply sled-agent/src/updates.rs]: I'll leave it to you; can you CC me on the review? I'd love to see.

**Pull request merged**